### DISCUSSION TITLE
Origin of haplotypes when using pullIbdHaplo()

### DISCUSSION THREAD

**Discussion by xcazenave** [gaynorr/AlphaSimR#192] [Q&A]
Hi everyone,

If I'l not mistaken, when using the `pullIbdHaplo()` function, the origin of the haplotypes in the studied population is given relative to the population of founders. Would there be a trick to obtain the origin of the haplotypes relative to another population?

For example, with the following code:

```
founderPop <- quickHaplo(nInd=2, nChr=1, segSites=4)

SP <- SimParam$new(founderPop)
SP$setTrackRec(TRUE)


pop <- newPop(founderPop, simParam=SP)
rand_pop <- randCross(pop, nCrosses = 1, nProgeny = 2)
rand_pop2 <- randCross(rand_pop, nCrosses = 1, nProgeny = 2)

pullIbdHaplo(rand_pop)

    1_1 1_2 1_3 1_4
3_1   1   2   2   2
3_2   3   3   3   4
4_1   2   1   1   1
4_2   4   4   3   4

pullIbdHaplo(rand_pop2)

    1_1 1_2 1_3 1_4
5_1   1   2   2   2
5_2   2   4   3   4
6_1   3   3   3   4
6_2   4   4   3   1
```
Would there be a way to obtain the origin of the haplotypes in the `rand_pop2` population relative to the `rand_pop` population? The result would then be:

```
pullIbdHaplo(rand_pop)

    1_1 1_2 1_3 1_4
1_1   1   1   1   1
1_2   2   2   2   2
2_1   3   3   3   3
2_2   4   4   4   4

pullIbdHaplo(rand_pop2)

    1_1 1_2 1_3 1_4
3_1   1   1   1   1
3_2   3   4   4   4
4_1   2   2   2   2
4_2   4   4   4   3
```
Thank you very much!

**Comment by gaynorr** [comment]
I don't have anything built-in to AlphaSimR for this, so you'd have to dig around a bit in the internals to work something out or use a hack.

Doing this by looking at the internals can take a while. You'll need two objects from SimParam: recHist (a history of observed recombinations) and pedigree.

The hack is much easier but may not work for you in all cases. The hack is to just use `newPop` to create a new population with rand_pop. The call to `newPop` will create of a new population with the exact same genotypes. Importantly, this new population will be treated as additional founders and added as new individuals to the pedigree. Here is an example using your code.

```
library(AlphaSimR)

founderPop <- quickHaplo(nInd=2, nChr=1, segSites=4)

SP <- SimParam$new(founderPop)
SP$setTrackRec(TRUE)


pop <- newPop(founderPop, simParam=SP)
rand_pop <- randCross(pop, nCrosses = 1, nProgeny = 2)

# Overwriting rand_pop using newPop to set rand_pop as additional founders
rand_pop = newPop(rand_pop)

rand_pop2 <- randCross(rand_pop, nCrosses = 1, nProgeny = 2)

pullIbdHaplo(rand_pop)

1_1 1_2 1_3 1_4
5_1   5   5   5   5
5_2   6   6   6   6
6_1   7   7   7   7
6_2   8   8   8   8

# Note that the number starts at 5, because the intial founders are still being counted
# In this example, you can just subtract by 4 to get what you want.

pullIbdHaplo(rand_pop) - 4

1_1 1_2 1_3 1_4
5_1   1   1   1   1
5_2   2   2   2   2
6_1   3   3   3   3
6_2   4   4   4   4

pullIbdHaplo(rand_pop2) - 4

1_1 1_2 1_3 1_4
7_1   2   1   2   2
7_2   3   4   4   4
8_1   1   1   2   2
8_2   4   4   4   4
```

**Reply by gregorgorjanc** [reply]
> The hack is much easier but may not work for you in all cases. The hack is to just use `newPop` to create a new population with rand_pop. The call to `newPop` will create of a new population with the exact same genotypes. Importantly, this new population will be treated as additional founders and added as new individuals to the pedigree. Here is an example using your code.

@gaynorr can I double check - a call to `newPop()` always adds individuals as founders? I think we should ensure this is clearly documented in the help page - at the moment it says this under Details:

```
Note that newPop takes genomes from the rawPop and uses them without recombination! Hence, if you call newPop(rawPop = founderGenomes) twice, you will get two sets of individuals with different id but the same genomes. To get genetically different sets of individuals you can subset the rawPop input, say first half for one set and the second half for the other set.
```

There is a note in the Description, but a cryptic one:

```
Creates an initial [Pop-class](http://127.0.0.1:47725/help/library/AlphaSimR/help/Pop-class) from an object of [MapPop-class](http://127.0.0.1:47725/help/library/AlphaSimR/help/MapPop-class) or [NamedMapPop-class](http://127.0.0.1:47725/help/library/AlphaSimR/help/NamedMapPop-class). The function is intended for use with output from functions such as [runMacs](http://127.0.0.1:47725/help/library/AlphaSimR/help/runMacs), [newMapPop](http://127.0.0.1:47725/help/library/AlphaSimR/help/newMapPop), or [quickHaplo](http://127.0.0.1:47725/help/library/AlphaSimR/help/quickHaplo).
```

Will provide a simple fix to the doc once you confirm;0

**Comment by xcazenave** [comment]
Thanks for the quick answer and the hack @gaynorr !
After reading it, I have another question: say that I want to get the origin of the haplotypes relative to `rand_pop` (as in the example above) but at the same time relative to the founders. I guess there is no way to "remove" the individuals of `rand_pop` from the founders, right? In this case, I guess I should use `recHist ` and the pedigree as you mentionned?

**Reply by gregorgorjanc** [reply]
> say that I want to get the origin of the haplotypes relative to `rand_pop` (as in the example above) but at the same time relative to the founders. I guess there is no way to "remove" the individuals of `rand_pop` from the founders, right? In this case, I guess I should use `recHist ` and the pedigree as you mentionned?

@xcazenave what is the motivation to do it relative twice - to the founder pop and to a non-founder pop - can you explain how you would do this even outside of AlphaSimR?
